File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <bits/stdc++.h>
using namespace std;

const int dy[] = {-1, 0, 1, 0};
const int dx[] = {0, 1, 0, -1};
int n, cnt, mx, a[104][104], visited[104][104];

void dfs(int y, int x, int h) {
  visited[y][x] = 1;

  for (int i = 0; i < 4; i++) {
    int ny = y + dy[i];
    int nx = x + dx[i];

    if (ny < 0 || nx < 0 || ny >= n || nx >= n || visited[ny][nx]) continue;
    if (a[ny][nx] <= h) continue;
    dfs(ny, nx, h);
  }
}

int main() {
  cin >> n;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      cin >> a[i][j];
    }
  }

  for (int h = 1; h <= 100; h++) {
    cnt = 0;
    memset(visited, 0, sizeof(visited));

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (a[i][j] <= h || visited[i][j]) continue;
        dfs(i, j, h);
        cnt++;
      }
    }
    mx = max(cnt, mx);
  }
  if (mx == 0) {
    cout << 1;
    return 0;
  }
  cout << mx;
  return 0;
}

File: 백준/Silver/2468. 안전 영역/안전 영역.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().split('\n');

const n = Number(input[0]);

const map = input.slice(1).map(line => line.split(' ').map(Number));

const dy = [-1, 0, 1, 0];
const dx = [0, 1, 0, -1];

const dfs = (y, x, h, visited) => {
    visited[y][x] = 1;

    for (let i=0; i<4; i++){
        const ny = y + dy[i];
        const nx = x + dx[i];

        if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
        if(map[ny][nx] <= h || visited[ny][nx]) continue;

        dfs(ny, nx, h, visited);
    }
}


const answer = (n, map) => {
    let ret = 0;

    for(let h=0; h<=100; h++){
        const visited = Array.from({length: n}, ()=>Array(n).fill(0));
        let cnt = 0;

        for (let i=0; i<n; i++){
            for (let j=0; j<n; j++){
                if(map[i][j] <= h || visited[i][j]) continue;
                dfs(i, j, h, visited);
                cnt++; 
            }
        }
        ret = Math.max(ret, cnt);
    }

    return ret;
}

console.log(answer(n, map));